cellpack/autopack/ingredient/Ingredient.py

```diff
@@ -2491,48 +2443,65 @@ def np_check_collision(self, position, rotation):
+        )
+        # if overlap_distance is negative, the encapsualting radii are overlapping
+        overlap_indices = numpy.nonzero(overlap_distance < 0.0)[0]
+
+        if len(overlap_indices) != 0:
+            level = level + 1
+            # for each packed ingredient that had a collision, we want to check the more
+            # detailed geometry, ie walk down the sphere tree file.
+            while level < total_levels:
+                pos_of_attempting_ingr = self.get_new_pos(
+                    self, packing_location, rotation, self.positions[level]
                 )
-                dist, ind = sph1.query(pos, len(p))
-                # return indice of sph1 closest to pos
-                cradii = numpy.take(self.radii[0], ind)
-                oradii = numpy.array(self.env.rIngr[indice].radii[0])
-                sumradii = numpy.add(cradii.transpose(), oradii).transpose()
-                sD = dist - sumradii
-                # sD = dist - (numpy.take(self.radii[0], ind)+numpy.array(self.env.rIngr[indice].radii[0]))
-                nbo = numpy.nonzero(sD < 0.0)[0]
-                # pdb.set_trace()
-                if len(nbo):
-                    overlap = True
-                    break
-            del sph1
+                search_tree_for_new_ingr = spatial.cKDTree(pos_of_attempting_ingr)
+                collision = False
+                # NOTE: At certain lengths of overlap_indices, it might help to remove items from the list
+                # if they dont have a collision at a non max level, but for short arrays, removing indices
+                # takes longer than not checking it.
+                for i in range(len(overlap_indices)):
+                    index = indices[overlap_indices[i]]
+                    collision = self.check_one_level_for_collision(
+                        index, level, search_tree_for_new_ingr
+                    )
+                    if collision:
+                        break
+                level += 1
+                if collision:
+                    if level == total_levels:
+                        # found collision at lowest level, break all the way out
+                        return True
+                del search_tree_for_new_ingr
+
         if self.compNum == 0:
             organelle = self.env
         else:
             organelle = self.env.compartments[abs(self.compNum) - 1]
         for o in self.env.compartments:
             if organelle.name == o.name:
                 continue
-            d, i = o.OGsrfPtsBht.query(position)
+            d, i = o.OGsrfPtsBht.query(packing_location)
             if d < self.encapsulatingRadius * 1.1:
-                p = self.get_new_pos(self, position, rotation)
-                d, i = o.OGsrfPtsBht.query(p)
-                D = d - numpy.array(self.radii[0])
-                nb = numpy.nonzero(D < 0.0)[0]
-                if len(nb) != 0:
+                pos_of_attempting_ingr = self.get_new_pos(
```

Comment:
We need to do the same changes here to use positions and rotations at a certain level instead of `[0]`

Reply:
Is this why we ended up getting some overlaps with the last iteration?

Reply:
no, I just figured out why the overlaps were happening: the radii of the packed ingr wasn't in the same order as the results from `query` so the distance check was sorta random. 

The last commit has these changes. 

![newplot (24)](https://user-images.githubusercontent.com/5170636/159086715-248ffddb-35fb-48b8-95c6-4392ca7c5677.png)
![newplot (25)](https://user-images.githubusercontent.com/5170636/159086722-eec7342e-fc40-4a17-9590-d422b50967ed.png)